### DISCUSSION TITLE
logged events suggest only one value of a variable may have linear dependence, how to solve this?

### DISCUSSION THREAD

**Discussion by ABQTrap** [amices/mice#610] [Q&A] [answered]
Dear all, I tried to impute missing values by MICE, there was warning message:
Number of logged events: 25 
`impute_training$loggedEvents`
   it im  dep meth          out
1   1  1 IGF1  pmm subject_ID59
2   1  2 IGF1  pmm subject_ID59
3   1  3 IGF1  pmm subject_ID59
4   1  4 IGF1  pmm subject_ID59
5   1  5 IGF1  pmm subject_ID59
6   2  1 IGF1  pmm subject_ID59
7   2  2 IGF1  pmm subject_ID59
8   2  3 IGF1  pmm subject_ID59
9   2  4 IGF1  pmm subject_ID59
10  2  5 IGF1  pmm subject_ID59
11  3  1 IGF1  pmm subject_ID59
12  3  2 IGF1  pmm subject_ID59
13  3  3 IGF1  pmm subject_ID59
14  3  4 IGF1  pmm subject_ID59
15  3  5 IGF1  pmm subject_ID59
16  4  1 IGF1  pmm subject_ID59
17  4  2 IGF1  pmm subject_ID59
18  4  3 IGF1  pmm subject_ID59
19  4  4 IGF1  pmm subject_ID59
20  4  5 IGF1  pmm subject_ID59
21  5  1 IGF1  pmm subject_ID59
22  5  2 IGF1  pmm subject_ID59
23  5  3 IGF1  pmm subject_ID59
24  5  4 IGF1  pmm subject_ID59
25  5  5 IGF1  pmm subject_ID59

the logged events suggested that the value of subject_ID59 of the variable subject_ID may have linear dependence problem, so did it mean that only this peticular value have collinear problem? If I don't want to exclude the whole variable subject_ID using predictor matrix, what should I do? Many thanks.

**Comment by stefvanbuuren** [comment, marked as answer]
You didn't say how you run your model, but did I suspect your variable `subject_ID` has class `character` or `factor`. In that case `mice` expands it into dummy variables. My advice is to keep `subject_ID` out of the imputation model. See below for a reprex and solution.

``` r
library(mice, warn.conflicts = FALSE)

# show problem
df <- mice::nhanes
df$subjid <- rownames(df)
imp1 <- mice(df, print = FALSE, seed = 1)
#> Warning: Number of logged events: 99
head(imp1$loggedEvents)
#>   it im dep     meth
#> 1  0  0     constant
#> 2  1  1 bmi      pmm
#> 3  1  1 bmi      pmm
#> 4  1  1 hyp      pmm
#> 5  1  1 hyp      pmm
#> 6  1  1 chl      pmm
#>                                                                                                        out
#> 1                                                                                                   subjid
#> 2                                                      df set to 1. # observed cases: 16  # predictors: 28
#> 3 age, hyp, subjid10, subjid11, subjid12, subjid13, subjid16, subjid21, subjid3, subjid4, subjid5, subjid6
#> 4                                                      df set to 1. # observed cases: 17  # predictors: 28
#> 5               age, bmi, chl, subjid10, subjid11, subjid12, subjid16, subjid2, subjid21, subjid4, subjid6
#> 6                                                      df set to 1. # observed cases: 15  # predictors: 28

# repair problem
pred <- make.predictorMatrix(df)
pred[, "subjid"] <- 0
imp2 <- mice(df, pred = pred, print = FALSE, seed = 1)
head(imp2$loggedEvents)
#> NULL
```

<sup>Created on 2023-12-27 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Reply by ABQTrap** [reply]
Dear stefvanbuuren, thank you for your reply. You are absolutely right, subject_ID has class of factor. I followed your advice, and logged events became NULL. I highly apprecitate your help.

**Reply by ABQTrap** [reply]
Dear stefvanbuuren, you mentioned that mice expands the variables with class of `factor` or `character` into dummy variables, does this mean that I should keep out other variables with class of `factor` out during imputing data, although logged events showed no warning message? I'm new to MICE, thanks a lot.

**Reply by stefvanbuuren** [reply]
No, keep any factors that are informative in the imputation model. But ID-variable is administrative, not related to the data, and has many different categories, so you better remove that one from the predictors (you still keep it in the data).
